otherlibs/systhreads/st_stubs.c

```diff
@@ -213,6 +216,7 @@ static void restore_runtime_state(caml_thread_t th)
   CAMLassert(th != NULL);
   Active_thread = th;
   Caml_state->current_stack = th->current_stack;
+  Caml_state->tls_state = th->tls_state;
```

Comment:
I think that since `Caml_state->tls_state` is registered as a generational global root, this should use `caml_modify_generational_global_root`.

Reply:
Question: shouldn't each `th->tls_state` field also be registered as a root?

Reply:
Answer: there is no need to register each `th->tls_state` as a root, because it is scanned during `caml_thread_scan_roots`. This also answers my question related to finalizers: it should be the case that the `tls_state` of a terminated thread is collected by the GC, as it will not be kept alive by `caml_thread_scan_roots` after thread termination.